# Линейная регрессия. Однофакторный дисперсионный анализ
### Задача 1

Даны значения величины заработной платы заемщиков банка `salary` и значения их поведенческого кредитного скоринга `scoring`:

In [1]:
salary = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110]
scoring = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]

Возьмём в качестве признака значение `salary`, а в качестве целевой переменной — `scoring`.

* Найдите коэффициенты линейной регрессии с помощью формул для парной регрессии, а затем с помощью метода наименьших квадратов.
* Постройте scatter plot по данным и отметьте на нём прямую линейной регрессии.
* Посчитайте коэффициент детерминации, среднюю ошибку аппроксимации.
* Оцените построенное уравнение регрессии с помощью F-критерия Фишера.
* Постройте для коэффициентов регрессии доверительные интервалы с помощью t-статистики Стьюдента.

#### Решение

In [2]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from scipy import stats

plt.style.use('seaborn-whitegrid')

In [3]:
x = np.array(salary)
y = np.array(scoring)

$y = b_0 + b x$

#### 1.1
Сначала найдем коэффициенты, используя формулы для парной регрессии:

$b = \dfrac{\overline{yx} - \overline{y} \cdot {\overline{x}}}{\overline{x^2} - {\overline{x}}^2}, \:\:
b_0 = \overline{y} - b {\overline{x}}.$

In [4]:
b = ((x * y).mean() - y.mean() * x.mean()) / ((x ** 2).mean() - x.mean() ** 2)

In [5]:
b0 = y.mean() - b * x.mean()

In [6]:
print(b0, b)

444.1773573243596 2.620538882402765


Получаем: $y = 444.18 + 2.62 \cdot x$

#### 1.2
Теперь посчитаем методом наименьших квадратов

$y = X \cdot b$

$(X \cdot b - y)^\top \cdot (X \cdot b - y) \rightarrow \min_b$

$b = (X^\top X)^{-1} \cdot X^\top \cdot y$

In [9]:
X = pd.DataFrame({'x0': np.ones(x.shape[0]),
                  'x1': x})
X.tail(3)

x0   x1
7  1.0  150
8  1.0  120
9  1.0  110

In [10]:
XTX = X.T.dot(X)
XTX_inv = np.linalg.inv(XTX)
B = XTX_inv.dot(X.T).dot(y)
print(*B)

444.17735732435915 2.6205388824027684


Аналогичный результат: $y = 444.18 + 2.62 \cdot x$

#### 2
Строим график

### Задача 2

Допустим, первые 5 клиентов из предыдущего задания проживают в Москве, а остальные — в Санкт-Петербурге. Влияет ли этот фактор на значение их кредитного скоринга?

#### Решение

> Вариационный анализ<br>
> Разделить выборку на две части<br>
> Нужен только таргет

### Задача 3

Посчитать коэффициенты линейной регрессии для т.н. Квартета Энскомба.

#### Решение